ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -48,32 +53,130 @@ private class PackageService(reloadPackageStoreIfChanged: PackageService.ReloadP
   @volatile private var state: State =
     State(Set.empty, TemplateIdMap.Empty, Map.empty, Map.empty, Map.empty)
 
-  // synchronized, so two threads cannot reload it concurrently
-  def reload(implicit
-      ec: ExecutionContext,
-      lc: LoggingContextOf[InstanceUUID],
-  ): Future[Error \/ Unit] =
-    synchronized {
-      reloadPackageStoreIfChanged(state.packageIds).map {
-        _.map {
-          case Some(diff) =>
-            this.state = this.state.append(diff)
-            logger.info(s"new package IDs loaded: ${diff.keySet.mkString(", ")}")
-            logger.debug(s"loaded diff: $diff")
-            ()
-          case None =>
-            logger.debug(s"new package IDs not found")
-            ()
+  // volatile, because we want to reduce the probability of unnecessary package updates
+  // due to multiple threads reading this.
+  // Writes to this are synchronized with a mutex, so in theory this should work out fine.
+  @volatile private var lastUpdated = LocalDateTime.MIN
+
+  // Regular updates should happen regardless of the current state every minute.
+  private def packagesShouldBeFetchedAgain =
+    lastUpdated.until(LocalDateTime.now(), temporal.ChronoUnit.SECONDS) >= 60L
+
+  private val _reload = {
+    // Because things can go so easily wrong, the declaration of the mutex
+    // was moved into a separate block, so it's easier to track down all interactions.
+    val packageUpdateSync = util.Concurrent.Mutex()
+    val updateTaskUpdateSync = util.Concurrent.Mutex()
+    @volatile var ongoingPackageUpdateTask = Future.successful(().right)
+    @volatile var isValid = false
+
+    def reload(jwt: Jwt)(implicit
+        ec: ExecutionContext,
+        lc: LoggingContextOf[InstanceUUID],
+    ): Future[Error \/ Unit] = {
+      def update() = {
+        val f =
+          EitherT
+            .eitherT(
+              Future(
+                logger.debug("Trying to execute a package update")
+              ) *> reloadPackageStoreIfChanged(
+                jwt
+              )(state.packageIds)
+            )
+            .map {
+              case Some(diff) =>
+                this.state = this.state.append(diff)
+                logger.info(s"new package IDs loaded: ${diff.keySet.mkString(", ")}")
+                logger.debug(s"loaded diff: $diff")
+              case None => logger.debug(s"new package IDs not found")
+            }
+            .map { res =>
+              lastUpdated = LocalDateTime.now()
+              res
+            }
+            .run
+        try {
+          ongoingPackageUpdateTask = f.map(_ => ().right)
+          // The task future has been updated, now all waiting threads can continue and pick it up.
+          isValid = true
```

Comment:
What is the subject of `isValid`? What is valid is `isValid` is `true`?

Reply:
I will remove that as it doesn't serve a purpose except for debugging